homeassistant/components/binary_sensor/mqtt.py

```diff
@@ -92,9 +105,23 @@ def async_added_to_hass(self):
         @callback
         def state_message_received(topic, payload, qos):
             """Handle a new received MQTT state message."""
+            if self._json_attributes:
+                self._attributes = {}
+                try:
+                    json_dict = json.loads(payload)
+                    if isinstance(json_dict, dict):
+                        attrs = {k: json_dict[k] for k in
+                                 self._json_attributes & json_dict.keys()}
```

Comment:
don't understand the use of `& json_dict.keys()`

Reply:
I makes sure you only get keys that exist in your defined `json_attributes` but also exist in json_dict. 

If you take out `json_dict.keys()` you will get an exception if there are `json_attributes` defined that didn't arrive in your message.

If you take out `json_dict.keys()` and received extra keys that were not defined in `json_attributes` you get those keys too.

Reply:
right!